File: packages/orchestrator/src/routing.ts
/**
 * Neuralgentics — Routing Matrix
 *
 * Hardcoded mapping from TaskType → Agent with forbidden agents.
 * CODE-LEVEL ENFORCED: the orchestrator MUST delegate based on these rules.
 */

import type { AgentRole, RoutingRule, TaskType } from './types.js';

export const ROUTING_MATRIX: RoutingRule[] = [
  {
    taskType: 'code-implementation',
    agent: 'coder',
    forbiddenAgents: ['explorer'],
    description: 'Writing/editing code, tests, config',
  },
  {
    taskType: 'architecture-design',
    agent: 'architect',
    forbiddenAgents: ['coder', 'explorer'],
    description: 'System design, trade-offs, research',
  },
  {
    taskType: 'file-finding',
    agent: 'explorer',
    forbiddenAgents: ['architect', 'coder', 'tester', 'writer', 'linter', 'git', 'scraper', 'mcp-specialist', 'release'],
    description: 'ONLY glob/find operations — no analysis',
  },
  {
    taskType: 'testing',
    agent: 'tester',
    forbiddenAgents: ['coder'],
    description: 'Test writing and test execution',
  },
  {
    taskType: 'linting',
    agent: 'linter',
    forbiddenAgents: ['coder', 'architect', 'tester'],
    description: 'Code style enforcement',
  },
  {
    taskType: 'git',
    agent: 'git',
    forbiddenAgents: ['coder', 'architect', 'tester', 'linter'],
    description: 'Commits, branches, tags',
  },
  {
    taskType: 'documentation',
    agent: 'writer',
    forbiddenAgents: ['explorer'],
    description: 'Markdown, README, docs',
  },
  {
    taskType: 'web-scraping',
    agent: 'scraper',
    forbiddenAgents: ['explorer'],
    description: 'URL fetching, data extraction',
  },
  {
    taskType: 'mcp-debug',
    agent: 'mcp-specialist',
    forbiddenAgents: ['explorer'],
    description: 'MCP protocol, server issues',
  },
  {
    taskType: 'release',
    agent: 'release',
    forbiddenAgents: ['coder', 'architect', 'tester', 'linter', 'explorer', 'scraper', 'writer', 'git', 'mcp-specialist'],
    description: 'Version bumps, changelogs',
  },
];

/** Lookup table for fast access: TaskType → RoutingRule */
const routingByType = new Map<TaskType, RoutingRule>(
  ROUTING_MATRIX.map((rule) => [rule.taskType, rule])
);

/**
 * Resolve a TaskType to the correct agent.
 * Returns null if no routing rule exists for the task type.
 */
export function resolveAgent(taskType: TaskType): AgentRole | null {
  const rule = routingByType.get(taskType);
  return rule?.agent ?? null;
}

/**
 * Check whether a given agent is forbidden for a task type.
 */
export function isForbiddenAgent(taskType: TaskType, agent: AgentRole): boolean {
  const rule = routingByType.get(taskType);
  if (!rule) return false;
  return rule.forbiddenAgents.includes(agent);
}

/**
 * Get the full routing rule for a task type.
 */
export function getRoutingRule(taskType: TaskType): RoutingRule | undefined {
  return routingByType.get(taskType);
}

/**
 * Validate that a task type → agent assignment is correct.
 * Returns { valid, violation? } object.
 */
export function validateRouting(taskType: TaskType, agent: AgentRole): {
  valid: boolean;
  expectedAgent: AgentRole | null;
  violation?: string;
} {
  const rule = routingByType.get(taskType);
  if (!rule) {
    return { valid: true, expectedAgent: null };
  }
  if (rule.agent !== agent) {
    return {
      valid: false,
      expectedAgent: rule.agent,
      violation: `ROUTING VIOLATION: ${taskType} should use ${rule.agent}, not ${agent}`,
    };
  }
  if (rule.forbiddenAgents.includes(agent)) {
    return {
      valid: false,
      expectedAgent: rule.agent,
      violation: `ROUTING VIOLATION: ${agent} is forbidden for ${taskType}`,
    };
  }
  return { valid: true, expectedAgent: rule.agent };
}

File: packages/orchestrator-go/src/neuralgentics/orchestrator/routing.go
package orchestrator

import "fmt"

// RoutingMatrix defines the CODE-LEVEL ENFORCED mapping from TaskType to AgentRole.
// The orchestrator MUST delegate based on these rules — no exceptions.
var RoutingMatrix = []RoutingRule{
	{
		TaskType:        TaskTypeCodeImpl,
		Agent:           AgentCoder,
		ForbiddenAgents: []AgentRole{AgentExplorer},
		Description:     "Writing/editing code, tests, config",
	},
	{
		TaskType:        TaskTypeArchDesign,
		Agent:           AgentArchitect,
		ForbiddenAgents: []AgentRole{AgentCoder, AgentExplorer},
		Description:     "System design, trade-offs, research",
	},
	{
		TaskType:        TaskTypeFileFinding,
		Agent:           AgentExplorer,
		ForbiddenAgents: []AgentRole{AgentArchitect, AgentCoder, AgentTester, AgentWriter, AgentLinter, AgentGit, AgentScraper, AgentMCPSpecialist, AgentRelease},
		Description:     "ONLY glob/find operations — no analysis",
	},
	{
		TaskType:        TaskTypeTesting,
		Agent:           AgentTester,
		ForbiddenAgents: []AgentRole{AgentCoder},
		Description:     "Test writing and test execution",
	},
	{
		TaskType:        TaskTypeLinting,
		Agent:           AgentLinter,
		ForbiddenAgents: []AgentRole{AgentCoder, AgentArchitect, AgentTester},
		Description:     "Code style enforcement",
	},
	{
		TaskType:        TaskTypeGit,
		Agent:           AgentGit,
		ForbiddenAgents: []AgentRole{AgentCoder, AgentArchitect, AgentTester, AgentLinter},
		Description:     "Commits, branches, tags",
	},
	{
		TaskType:        TaskTypeDocumentation,
		Agent:           AgentWriter,
		ForbiddenAgents: []AgentRole{AgentExplorer},
		Description:     "Markdown, README, docs",
	},
	{
		TaskType:        TaskTypeWebScraping,
		Agent:           AgentScraper,
		ForbiddenAgents: []AgentRole{AgentExplorer},
		Description:     "URL fetching, data extraction",
	},
	{
		TaskType:        TaskTypeMCPDebug,
		Agent:           AgentMCPSpecialist,
		ForbiddenAgents: []AgentRole{AgentExplorer},
		Description:     "MCP protocol, server issues",
	},
	{
		TaskType:        TaskTypeRelease,
		Agent:           AgentRelease,
		ForbiddenAgents: []AgentRole{AgentCoder, AgentArchitect, AgentTester, AgentLinter, AgentExplorer, AgentScraper, AgentWriter, AgentGit, AgentMCPSpecialist},
		Description:     "Version bumps, changelogs",
	},
}

// routingByType is the lookup table for O(1) TaskType → RoutingRule resolution.
var routingByType map[TaskType]RoutingRule

func init() {
	routingByType = make(map[TaskType]RoutingRule, len(RoutingMatrix))
	for _, rule := range RoutingMatrix {
		routingByType[rule.TaskType] = rule
	}
}

// ResolveAgent returns the designated AgentRole for a given TaskType.
// Returns an error if no routing rule exists for the task type and
// noDefault is true.
func ResolveAgent(taskType TaskType) (AgentRole, error) {
	rule, ok := routingByType[taskType]
	if !ok {
		return "", fmt.Errorf("no routing rule for task type: %s", taskType)
	}
	return rule.Agent, nil
}

// IsForbiddenAgent checks whether a given agent is forbidden for a task type.
func IsForbiddenAgent(taskType TaskType, agent AgentRole) bool {
	rule, ok := routingByType[taskType]
	if !ok {
		return false
	}
	for _, forbidden := range rule.ForbiddenAgents {
		if forbidden == agent {
			return true
		}
	}
	return false
}

// GetRoutingRule returns the full routing rule for a task type, or nil.
func GetRoutingRule(taskType TaskType) *RoutingRule {
	rule, ok := routingByType[taskType]
	if !ok {
		return nil
	}
	return &rule
}

// ValidateRouting validates that a task type → agent assignment is correct.
// Returns a RoutingValidation with valid=true if the routing is correct,
// or valid=false with a violation message if it's not.
func ValidateRouting(taskType TaskType, agent AgentRole) RoutingValidation {
	rule, ok := routingByType[taskType]
	if !ok {
		// No routing rule defined — allow any agent
		return RoutingValidation{Valid: true, ExpectedAgent: ""}
	}

	if rule.Agent != agent {
		return RoutingValidation{
			Valid:         false,
			ExpectedAgent: rule.Agent,
			Violation:     fmt.Sprintf("ROUTING VIOLATION: %s should use %s, not %s", taskType, rule.Agent, agent),
		}
	}

	for _, forbidden := range rule.ForbiddenAgents {
		if forbidden == agent {
			return RoutingValidation{
				Valid:         false,
				ExpectedAgent: rule.Agent,
				Violation:     fmt.Sprintf("ROUTING VIOLATION: %s is forbidden for %s", agent, taskType),
			}
		}
	}

	return RoutingValidation{Valid: true, ExpectedAgent: rule.Agent}
}
